Issue 315: Added PDFs as enclosures (attachments) in RSS feed

voutilad opened the pull request:
Per [Issue 315](https://github.com/freelawproject/courtlistener/issues/315), I've made a first cut of adding enclosures support to the RSS for Jurisdictions and the Search RSS feeds. 

Quite a few changes/additions, mostly lots of tests. Here's an example of the new output (with summary snipped out). The enclosure link appears after the summary element towards the end of the entry.

```xml
<?xml version="1.0" encoding="utf-8"?>
<feed xmlns="http://www.w3.org/2005/Atom" xml:lang="en-us">
	<title>CourtListener.com: All opinions for the Supreme Court of Vermont</title>
	<link href="https://www.courtlistener.com" rel="alternate"/>
	<link href="http://example.com/feed/court/vt/" rel="self"/>
	<id>https://www.courtlistener.com</id>
	<updated>2015-12-24T00:00:00-08:00</updated>
	<author>
		<name>Free Law Project</name>
		<email>feeds@courtlistener.com</email>
	</author>
	<rights>Created for the public domain by Free Law Project</rights>
	<entry>
		<title>Vermont Human Rights Commission, Lynne Silloway, Mary Bertrand and Lisa DeBlois v. State of Vermont, Department of Corrections and Department of Human Services</title>
		<link href="http://example.com/opinion/13/vermont-human-rights-commission-lynne-silloway-mary-bertrand-and-lisa/" rel="alternate"/>
		<published>2015-12-24T00:00:00-08:00</published>
		<author>
			<name>Supreme Court of Vermont</name>
		</author>
		<id>http://example.com/opinion/13/vermont-human-rights-commission-lynne-silloway-mary-bertrand-and-lisa/</id>
		<summary type="html">...</summary>
		<link length="218487" href="/pdf/2015/12/24/vermont_human_rights_commission_lynne_silloway_mary_bertrand_and_lisa.pdf" type="application/pdf" rel="enclosure"/>
		<category term="Precedential"/>
	</entry>
...
```

Core logic changes are in `cl.search.feeds.JurisdictionFeed` adding the methods needed for Django to build out the enclosures. I did have to update/fix the `cl.simple_pages.views.serve_static_file` function to use the API now in the `opinion-split` branch (i.e. add support for `Opinion` models).

Tests include functional (added [FeedParser](https://pythonhosted.org/feedparser/) to formally test parsing of RSS feeds) and some unit tests. Also, to support some of the tests I had to add some binary objects into the source tree. Current design is to mimic the pattern of the production asset directory (starting with file extension then following date segments) and the tests use an `@override_settings` to rebase the test asset directory on a per test basis:

```python
@override_settings(
    MEDIA_ROOT = os.path.join(settings.INSTALL_ROOT, 'cl/assets/media/test/')
)
class TestClassName(TestCase):
...
```

Since the `JurisdictionFeed` originally entirely relied on the Solr index for generating the feed, it makes sense to potentially add the key metadata needed to the index to prevent repeated disk hits for doing things like file size calculation and mime-type generation.

mlissner commented on cl/search/feeds.py:
```diff
@@ -97,6 +97,38 @@ def item_title(self, item):
+        try:
+            path = item['local_path']
+            if not path.startswith('/'):
+                return '/%s' % (path,)
```
This is weird. What's up with this?

voutilad replied:
This is the Django Feed object will render it as path off the root of the site vs. a relative path to the feed URL. Without it, you end up with something like `https://www.courtlistener.com/feeds/court/scotus/pdf/2015/12/30/filename.pdf`
instead of
`https://www.courtlistener.com/pdf/2015/12/30/filename.pdf`

mlissner replied:
Gotcha. This actually makes me think that we should use `reverse` here too, reversing the URL for `serve` static file. That should fix this issue, making it explicit how it works, and also allow us to change the URL down the road if we need to.

mlissner commented on cl/search/feeds.py:
```diff
@@ -97,6 +97,38 @@ def item_title(self, item):
+            mime = magic.from_file(file_loc, mime=True)
+            if mime.startswith('inode'):
+                return None
+            return mime
```
We use `magic` for this [in our crawler][1] and have identified a lot of weird things over the years, with their resulting workarounds. I wonder if it's better to reverse the extension (which incorporates the workarounds) to get the mimetype?

[1]: https://github.com/freelawproject/courtlistener/blob/opinion-split/cl/lib/scrape_helpers.py#L48-L77

voutilad replied:
Easily doable and I think the only consequence being if it's for some accurate, it should only cause a minor hiccup on how the client tries to handle the response, but shouldn't block the download. I'll assume that:

* There's a file extension and it's accurate.
* If not I'll check the parent directory (typically 'pdf/') and assume that's accurate.
* All else fails, default to `application/octet-stream` so the client should properly see it as a downloadable binary file and not try to render the byte stream. (Only side effect could be a text-based file slipping through mime-type guessing and ending up downloaded and not displayed natively in the browser.)

Thinking it best to use the mimetypes based on file extensions as shown on this [StackOverflow](http://stackoverflow.com/questions/4212861/what-is-a-correct-mime-type-for-docx-pptx-etc) post.

mlissner replied:
The parent directory will be the same as the extension. They're set the same way, so if there's no extension at all, feel free to go straight to application/octet-stream.

mlissner commented on cl/simple_pages/tests.py:
```diff
@@ -23,3 +30,77 @@ def test_robots_page(self):
+        self.audio.save(index=False)
+
+        self.opinioncluster = OpinionCluster(
+            case_name=u'Hotline Bling',
```
Nice. Well played. Whatever you do, don't listen to the various testing files that are already in the code base.

mlissner commented on cl/tests/test_feeds.py:
```diff
@@ -0,0 +1,154 @@
+        Does the feeds page show all the proper links for each jurisdiction?
+        """
+        courts = Court.objects.all()
+        self.browser.get('%s/feeds' % (self.server_url,))
```
This should use [reverse][1] instead. Ideally we shouldn't ever hard code URLs anywhere other than the urls.py file. This one, for example can be replaced with:

    '%s%s' % (self.server_url, reverse('feed_info'))

You can get a list of all the URLs and their names by using something like:

    manage.py show_urls | grep feeds

(Though this is only available if you have `django-extensions` installed.)


[1]: https://docs.djangoproject.com/en/1.8/ref/urlresolvers/#reverse

mlissner commented on cl/tests/test_feeds.py:
```diff
@@ -0,0 +1,154 @@
+            link = self.browser.find_element_by_link_text(court.full_name)
+            self.assertEqual(
+                link.get_attribute('href'),
+                '%s/feed/court/%s/' % (self.server_url, court.pk,)
```
Ditto here. This one needs a name added to it though in `urls.py` (it lacks a name presently, alas), but then can be called with something like:

    reverse('jurisdiction-feed', kwargs={'court': court.pk})

mlissner commented:
All in all, looks great. I'll pull and make the tweaks that I commented on above. I am curious though about the need to add a slash some of the time (my first comment).

voutilad commented:
Seems this is still open. Let me know if you want me to update with the tweaks you mentioned.

mlissner commented:
Yeah, if you don't mind making these tweaks, that'd be great. I meant to get back to it, but I've barely gotten anything done this week.

voutilad commented:
Thought of a cute way to work around the bug I reported in Django without having to deal with changing any production logic at the moment: override the `path` method in the test class before the `setUp()` logic touches the offending indexing code. See [f498927](https://github.com/voutilad/courtlistener/commit/f4989279ec998ec5affba1cf763aff55631791a7)

Working on fixing the logic to remove "magic" and use "reverse" next.

voutilad commented:
Ok, I think I've incorporated what you were looking for above. I put together a (currently overkill) lookup dictionary of mimetypes based on file extensions and swapped out the `magic` usage. Also refactored the tests to use `reverse()` calls and in the process had to add in names to some of the routes/url pattern entries for the feeds. Let me know if something is missing, but looks done. (At least the tests I wrote are passing again/still.)

mlissner commented:
Very nice stuff.
